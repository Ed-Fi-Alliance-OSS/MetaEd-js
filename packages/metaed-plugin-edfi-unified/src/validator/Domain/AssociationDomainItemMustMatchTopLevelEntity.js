// @flow
import type { DomainItem, MetaEdEnvironment, ValidationFailure } from '../../../../../packages/metaed-core/index';

function getFailure(domainItem: DomainItem, name: string, failureMessage: string): ValidationFailure {
  return {
    validatorName: name,
    category: 'error',
    message: failureMessage,
    sourceMap: domainItem.sourceMap.type,
    fileMap: null,
  };
}

// eslint-disable-next-line no-unused-vars
export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.entity.domain.forEach(domain => {
    domain.domainItems.forEach(domainItem => {
      if (domainItem.referencedType === 'association') {
        if (!metaEd.entity.association.has(domainItem.metaEdName) && !metaEd.entity.associationSubclass.has(domainItem.metaEdName)) {
          failures.push(getFailure(domainItem, 'AssociationDomainItemMustMatchTopLevelEntity',
            `Association Domain Item property '${domainItem.metaEdName}' does not match any declared Association or Association Subclass.`));
        }
      }
    });
  });

  return failures;
}
