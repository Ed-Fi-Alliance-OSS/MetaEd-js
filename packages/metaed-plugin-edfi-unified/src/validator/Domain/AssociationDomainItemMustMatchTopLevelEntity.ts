import { DomainItem, MetaEdEnvironment, ValidationFailure, Namespace } from 'metaed-core';
import { getEntityFromNamespaceChain } from 'metaed-core';

function getFailure(domainItem: DomainItem, name: string, failureMessage: string): ValidationFailure {
  return {
    validatorName: name,
    category: 'error',
    message: failureMessage,
    sourceMap: domainItem.sourceMap.metaEdName,
    fileMap: null,
  };
}

export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];
  metaEd.namespace.forEach((namespace: Namespace) => {
    namespace.entity.domain.forEach(domain => {
      domain.domainItems.forEach(domainItem => {
        if (domainItem.referencedType !== 'association') return;
        if (
          getEntityFromNamespaceChain(
            domainItem.metaEdName,
            domainItem.referencedNamespaceName,
            namespace,
            'association',
            'associationSubclass',
          ) == null
        ) {
          failures.push(
            getFailure(
              domainItem,
              'AssociationDomainItemMustMatchTopLevelEntity',
              `Association Domain Item property '${
                domainItem.metaEdName
              }' does not match any declared Association or Association Subclass in namespace ${
                domainItem.referencedNamespaceName
              }.`,
            ),
          );
        }
      });
    });
  });

  return failures;
}
