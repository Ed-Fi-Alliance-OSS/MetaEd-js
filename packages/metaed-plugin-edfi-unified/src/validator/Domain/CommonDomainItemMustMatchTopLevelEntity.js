// @flow
import type { DomainItem, MetaEdEnvironment, ValidationFailure } from '../../../../metaed-core/index';

function getFailure(domainItem: DomainItem, name: string, failureMessage: string): ValidationFailure {
  return {
    validatorName: name,
    category: 'error',
    message: failureMessage,
    sourceMap: domainItem.sourceMap.type,
    fileMap: null,
  };
}

// eslint-disable-next-line no-unused-vars
export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.entity.domain.forEach(domain => {
    domain.domainItems.forEach(domainItem => {
      if (domainItem.referencedType === 'common') {
        if (!metaEd.entity.common.has(domainItem.metaEdName)) {
          failures.push(getFailure(domainItem, 'CommonDomainItemMustMatchTopLevelEntity',
            `Common Domain Item property '${domainItem.metaEdName}' does not match any declared Common.`));
        }
      }
    });
  });

  return failures;
}
