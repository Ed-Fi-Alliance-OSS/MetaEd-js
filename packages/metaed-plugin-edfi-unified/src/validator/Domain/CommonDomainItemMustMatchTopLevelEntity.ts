import { DomainItem, MetaEdEnvironment, ValidationFailure, Namespace } from 'metaed-core';
import { getEntityFromNamespaceChain } from 'metaed-core';

function getFailure(domainItem: DomainItem, name: string, failureMessage: string): ValidationFailure {
  return {
    validatorName: name,
    category: 'error',
    message: failureMessage,
    sourceMap: domainItem.sourceMap.metaEdName,
    fileMap: null,
  };
}

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.namespace.forEach((namespace: Namespace) => {
    namespace.entity.domain.forEach(domain => {
      domain.domainItems.forEach(domainItem => {
        if (domainItem.referencedType !== 'common') return;
        if (
          getEntityFromNamespaceChain(domainItem.metaEdName, domainItem.referencedNamespaceName, namespace, 'common') == null
        ) {
          failures.push(
            getFailure(
              domainItem,
              'CommonDomainItemMustMatchTopLevelEntity',
              `Common Domain Item property '${domainItem.metaEdName}' does not match any declared Common in namespace ${
                domainItem.referencedNamespaceName
              }.`,
            ),
          );
        }
      });
    });
  });

  return failures;
}
