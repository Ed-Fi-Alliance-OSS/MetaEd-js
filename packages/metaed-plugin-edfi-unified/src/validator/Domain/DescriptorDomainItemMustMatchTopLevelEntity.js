// @flow
import type { DomainItem, MetaEdEnvironment, ValidationFailure } from 'metaed-core';

function getFailure(domainItem: DomainItem, name: string, failureMessage: string): ValidationFailure {
  return {
    validatorName: name,
    category: 'error',
    message: failureMessage,
    sourceMap: domainItem.sourceMap.type,
    fileMap: null,
  };
}

// eslint-disable-next-line no-unused-vars
export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.entity.domain.forEach(domain => {
    domain.domainItems.forEach(domainItem => {
      if (domainItem.referencedType === 'descriptor') {
        if (!metaEd.entity.descriptor.has(domainItem.metaEdName)) {
          failures.push(
            getFailure(
              domainItem,
              'DescriptorDomainItemMustMatchTopLevelEntity',
              `Descriptor Domain Item property '${domainItem.metaEdName}' does not match any declared Descriptor.`,
            ),
          );
        }
      }
    });
  });

  return failures;
}
