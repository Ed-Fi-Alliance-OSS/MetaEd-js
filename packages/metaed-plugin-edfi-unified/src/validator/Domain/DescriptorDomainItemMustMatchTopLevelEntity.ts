import { DomainItem, MetaEdEnvironment, ValidationFailure, Namespace } from 'metaed-core';
import { getEntityFromNamespaceChain } from 'metaed-core';

function getFailure(domainItem: DomainItem, name: string, failureMessage: string): ValidationFailure {
  return {
    validatorName: name,
    category: 'error',
    message: failureMessage,
    sourceMap: domainItem.sourceMap.metaEdName,
    fileMap: null,
  };
}

export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];

  metaEd.namespace.forEach((namespace: Namespace) => {
    namespace.entity.domain.forEach(domain => {
      domain.domainItems.forEach(domainItem => {
        if (domainItem.referencedType !== 'descriptor') return;
        if (
          getEntityFromNamespaceChain(domainItem.metaEdName, domainItem.referencedNamespaceName, namespace, 'descriptor') ==
          null
        ) {
          failures.push(
            getFailure(
              domainItem,
              'DescriptorDomainItemMustMatchTopLevelEntity',
              `Descriptor Domain Item property '${
                domainItem.metaEdName
              }' does not match any declared Descriptor in namespace ${domainItem.referencedNamespaceName}.`,
            ),
          );
        }
      });
    });
  });

  return failures;
}
