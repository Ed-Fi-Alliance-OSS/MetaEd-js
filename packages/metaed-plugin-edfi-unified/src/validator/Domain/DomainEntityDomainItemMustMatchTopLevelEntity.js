// @flow
import type { DomainItem, MetaEdEnvironment, ValidationFailure } from '../../../../../packages/metaed-core/index';

function getFailure(domainItem: DomainItem, name: string, failureMessage: string): ValidationFailure {
  return {
    validatorName: name,
    category: 'error',
    message: failureMessage,
    sourceMap: domainItem.sourceMap.type,
    fileMap: null,
  };
}

// eslint-disable-next-line no-unused-vars
export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.entity.domain.forEach(domain => {
    domain.domainItems.forEach(domainItem => {
      if (domainItem.referencedType === 'domainEntity') {
        if (!metaEd.entity.domainEntity.has(domainItem.metaEdName) && !metaEd.entity.domainEntitySubclass.has(domainItem.metaEdName)) {
          failures.push(getFailure(domainItem, 'DomainEntityDomainItemMustMatchTopLevelEntity',
            `Domain Entity Domain Item property '${domainItem.metaEdName}' does not match any declared Domain Entity or Domain Entity Subclass.`));
        }
      }
    });
  });

  return failures;
}
