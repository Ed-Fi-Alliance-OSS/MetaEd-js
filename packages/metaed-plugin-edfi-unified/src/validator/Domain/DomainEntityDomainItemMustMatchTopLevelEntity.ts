import { DomainItem, MetaEdEnvironment, ValidationFailure, Namespace } from 'metaed-core';
import { getEntityFromNamespaceChain } from 'metaed-core';

function getFailure(domainItem: DomainItem, name: string, failureMessage: string): ValidationFailure {
  return {
    validatorName: name,
    category: 'error',
    message: failureMessage,
    sourceMap: domainItem.sourceMap.referencedType,
    fileMap: null,
  };
}

export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];

  metaEd.namespace.forEach((namespace: Namespace) => {
    namespace.entity.domain.forEach(domain => {
      domain.domainItems.forEach(domainItem => {
        if (domainItem.referencedType !== 'domainEntity') return;
        if (
          getEntityFromNamespaceChain(
            domainItem.metaEdName,
            domainItem.referencedNamespaceName,
            namespace,
            'domainEntity',
            'domainEntitySubclass',
          ) == null
        ) {
          failures.push(
            getFailure(
              domainItem,
              'DomainEntityDomainItemMustMatchTopLevelEntity',
              `Domain Entity Domain Item property '${domainItem.metaEdName}' does not match any declared Domain Entity or Domain Entity Subclass in namespace ${domainItem.referencedNamespaceName}.`,
            ),
          );
        }
      });
    });
  });

  return failures;
}
