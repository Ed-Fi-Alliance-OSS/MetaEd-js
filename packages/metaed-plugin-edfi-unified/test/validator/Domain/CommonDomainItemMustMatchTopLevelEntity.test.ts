// SPDX-License-Identifier: Apache-2.0
// Licensed to the Ed-Fi Alliance under one or more agreements.
// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
// See the LICENSE and NOTICES files in the project root for more information.

import {
  newMetaEdEnvironment,
  MetaEdTextBuilder,
  DomainBuilder,
  CommonBuilder,
  NamespaceBuilder,
  CommonSubclassBuilder,
} from '@edfi/metaed-core';
import { MetaEdEnvironment, ValidationFailure } from '@edfi/metaed-core';
import { validate } from '../../../src/validator/Domain/CommonDomainItemMustMatchTopLevelEntity';

describe('when validating common domain item matches top level entity', (): void => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const domainName = 'DomainName';
  const commonName = 'CommonName';

  let failures: ValidationFailure[];
  let coreNamespace: any = null;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartDomain(domainName)
      .withDocumentation('doc')
      .withCommonDomainItem(commonName)
      .withFooterDocumentation('FooterDocumentation')
      .withEndDomain()

      .withStartCommon(commonName)
      .withDocumentation('doc')
      .withBooleanProperty('PropertyName', 'doc', true, false)
      .withEndCommon()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainBuilder(metaEd, []))
      .sendToListener(new CommonBuilder(metaEd, []));

    coreNamespace = metaEd.namespace.get('EdFi');
    failures = validate(metaEd);
  });

  it('should build one domain', (): void => {
    expect(coreNamespace.entity.domain.size).toBe(1);
  });

  it('should have no validation failures()', (): void => {
    expect(failures).toHaveLength(0);
  });
});

describe('when validating common domain item matches subclass', (): void => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const domainName = 'DomainName';
  const commonName = 'CommonName';

  let failures: ValidationFailure[];
  let coreNamespace: any = null;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartDomain(domainName)
      .withDocumentation('doc')
      .withCommonDomainItem(commonName)
      .withFooterDocumentation('FooterDocumentation')
      .withEndDomain()

      .withStartCommonSubclass(commonName, 'BaseName')
      .withDocumentation('doc')
      .withBooleanProperty('PropertyName', 'doc', true, false)
      .withEndCommonSubclass()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainBuilder(metaEd, []))
      .sendToListener(new CommonSubclassBuilder(metaEd, []));

    coreNamespace = metaEd.namespace.get('EdFi');
    failures = validate(metaEd);
  });

  it('should build one domain', (): void => {
    expect(coreNamespace.entity.domain.size).toBe(1);
  });

  it('should have no validation failures()', (): void => {
    expect(failures).toHaveLength(0);
  });
});

describe('when validating common domain item matches top level entity across namespace', (): void => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const domainName = 'DomainName';
  const commonName = 'CommonName';

  let failures: ValidationFailure[];
  let extensionNamespace: any = null;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartCommon(commonName)
      .withDocumentation('doc')
      .withBooleanProperty('PropertyName', 'doc', true, false)
      .withEndCommon()
      .withEndNamespace()

      .withBeginNamespace('Extension')
      .withStartDomain(domainName)
      .withDocumentation('doc')
      .withCommonDomainItem(`EdFi.${commonName}`)
      .withFooterDocumentation('FooterDocumentation')
      .withEndDomain()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainBuilder(metaEd, []))
      .sendToListener(new CommonBuilder(metaEd, []));

    const coreNamespace = metaEd.namespace.get('EdFi');
    extensionNamespace = metaEd.namespace.get('Extension');
    extensionNamespace.dependencies = [coreNamespace];
    failures = validate(metaEd);
  });

  it('should build one domain entity', (): void => {
    expect(extensionNamespace.entity.domain.size).toBe(1);
  });

  it('should have no validation failures()', (): void => {
    expect(failures).toHaveLength(0);
  });
});

describe('when validating common domain item does not match top level entity', (): void => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const domainName = 'DomainName';
  const commonName = 'CommonName';

  let failures: ValidationFailure[];
  let coreNamespace: any = null;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartDomain(domainName)
      .withDocumentation('doc')
      .withCommonDomainItem('CommonDomainItemName')
      .withFooterDocumentation('FooterDocumentation')
      .withEndDomain()

      .withStartCommon(commonName)
      .withDocumentation('doc')
      .withBooleanProperty('PropertyName', 'doc', true, false)
      .withEndCommon()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainBuilder(metaEd, []))
      .sendToListener(new CommonBuilder(metaEd, []));

    coreNamespace = metaEd.namespace.get('EdFi');
    failures = validate(metaEd);
  });

  it('should build one domain entity', (): void => {
    expect(coreNamespace.entity.domain.size).toBe(1);
  });

  it('should have one validation failure()', (): void => {
    expect(failures).toHaveLength(1);
    expect(failures[0].validatorName).toBe('CommonDomainItemMustMatchTopLevelEntity');
    expect(failures[0].category).toBe('error');
    expect(failures[0].message).toMatchSnapshot();
    expect(failures[0].sourceMap).toMatchSnapshot();
  });
});
