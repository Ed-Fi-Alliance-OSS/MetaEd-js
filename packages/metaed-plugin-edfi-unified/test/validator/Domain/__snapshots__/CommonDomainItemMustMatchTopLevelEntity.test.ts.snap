// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when validating common domain item does not match top level entity should have one validation failure() 1`] = `"Common Domain Item property 'CommonDomainItemName' does not match any declared Common in namespace EdFi."`;

exports[`when validating common domain item does not match top level entity should have one validation failure() 2`] = `
Object {
  "column": 11,
  "line": 5,
  "tokenText": "CommonDomainItemName",
}
`;
